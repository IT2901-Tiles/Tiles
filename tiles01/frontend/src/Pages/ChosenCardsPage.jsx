import React from 'react'
import LinkButton from '../Components/LinkButton'
import placeholderImage from '../images/blankImg.png'
import TopContainer from '../Components/TopContainer'
import LinkToPage from '../Components/LinkToPage'
import { useHistory } from "react-router-dom";
import { useState } from 'react';

function ChosenCardsPage(props) {

    try {
        localStorage.setItem(props.location.state.category, props.location.state.card)
    } catch (error) {
        console.log("no cards chosen")
    }

    var triggerChosen = !["null", null].includes(localStorage.getItem("trigger"))
    var thingChosen = !["null", null].includes(localStorage.getItem("things"))
    var feedbackChosen = !["null", null].includes(localStorage.getItem("feedback"))

    let triggerNotChosenId = "textNoCardChosen1";
    let thingNotChosenId = "textNoCardChosen2";
    let feedbackNotChosenId = "textNoCardChosen3";
    let triggerChosenId = "trigger";
    let thingChosenId = "things";
    let feedbackChosenId = "feedback";
    let errorOccured = false; // value turns true if error occured 
    let colorBorder; // color of border and font when no card is picked
    const [buttonActive, setButtonActive] = useState(triggerChosen && thingChosen && feedbackChosen);
    // This function checks which cards are checked off and sets a bool value


    // If bool value is true, then this function is returned. This returns the image and uses the LinkToPage-component
    function trueValue(cardType) {
        if (errorOccured) { // error-solution for now, this will be fixed more professionally later on
            return <LinkButton target="/categories" title={cardType.charAt(0).toUpperCase + cardType.substring(1)} category={cardType} size="Small"></LinkButton>
        } else {
            let cardTypeName = cardType.charAt(0).toUpperCase() + cardType.substring(1)
            return <LinkToPage target="/categories" title={cardTypeName} category={cardType} nameOfCardType={localStorage.getItem(cardType)} classNameCard={"chosen" + cardTypeName}></LinkToPage>
        }
    }

    // If bool value is false, then this function is returned. This function returns a placeholder image and a button
    function falseValue(cardType) {
        if (errorOccured) { // error-solution for now, this will be fixed
            return <LinkButton target="/categories" title={cardType} category={cardType} size="Small"></LinkButton>
        } else {
            // Checks which type of card it is + sets the color of the border
            let classNameOfCard;
            if (cardType === triggerChosenId) {
                classNameOfCard = triggerNotChosenId;
                colorBorder = "#F08A00";
            } else if (cardType === thingChosenId) {
                classNameOfCard = thingNotChosenId;
                colorBorder = "#E53D2F";
            } else if (cardType === feedbackChosenId) {
                classNameOfCard = feedbackNotChosenId;
                colorBorder = "#FFCD00";
            }

            // Returns the placeholder image and button
            // (fjern kommentar: kanskje litt dårlig navngivning på divene og kanskje unødvendig mange, men ble finest og mest responsivt med 2 diver)
            return <div id="firstWrapperCardNotChosen"><div className="wrapperCardNotChosen">
                <img src={placeholderImage} alt="test" className="noImgChosen" style={{ border: `${colorBorder}`, borderStyle: "solid", borderWidth: "0.65vw" }} />
                <div id={classNameOfCard} style={{ color: `${colorBorder}` }}>
                    {cardType.charAt(0).toUpperCase() + cardType.substring(1)}
                </div>
                <br />
            </div>
                <div id="buttonChosenCardsPage">
                    {/* Following code decides whether button is going to be enabled or not */}
                    {((cardType === triggerChosenId) && (!(thingChosen)) && (!(feedbackChosen))) ? <LinkButton target="/categories" title="Choose!" category="trigger" size="chosenCards" className="buttonChosenCardsPage"></LinkButton> : ""}
                    {((cardType === thingChosenId) && (triggerChosen) && (!(feedbackChosen))) ? <LinkButton target="/categories" title="Choose!" category="things" size="chosenCards" className="buttonChosenCardsPage"></LinkButton> : ""}
                    {((cardType === feedbackChosenId) && (triggerChosen) && (thingChosen)) ? <LinkButton target="/categories" title="Choose!" category="feedback" size="chosenCards" className="buttonChosenCardsPage"></LinkButton> : ""}

                </div>
            </div>

        }
    }
    return (
        <div className="Page">
            <TopContainer
                text="Some text"
            />

            <div className="chosenCardsGrid">
                <div className="chosenCardsGrid2">

                    {/* If-else statements to check the bool-variables */}
                    {(triggerChosen) ? <div className="chosenCardContainer">{trueValue(triggerChosenId)}</div> : falseValue(triggerChosenId)}
                    {(thingChosen) ? <div className="chosenCardContainer">{trueValue(thingChosenId)}</div> : falseValue(thingChosenId)}
                    {(feedbackChosen) ? <div className="chosenCardContainer">{trueValue(feedbackChosenId)}</div> : falseValue(feedbackChosenId)}
                </div>
            </div>



            <br />
            <LinkButton target="/animation" title="RUN!" category="animation" size="Large" active={buttonActive} buttonError={"Please select one card in each category"}   ></LinkButton>
        </div>
    );
}
//checks if cards have been selected


export default ChosenCardsPage