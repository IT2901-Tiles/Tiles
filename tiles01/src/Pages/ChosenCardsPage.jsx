import React from 'react'
import LinkButton from '../Components/LinkButton'
import placeholderImage from '../images/blankImg.png'
import TopContainer from '../Components/TopContainer'
import LinkToPage from '../Components/LinkToPage'

function ChosenCardsPage(props) {
    localStorage.setItem(props.location.state.category,props.location.state.card)
    let triggerChosen = false;
    let thingChosen = false;
    let feedbackChosen = false;
    let triggerNotChosenId = "textNoCardChosen1";
    let thingNotChosenId = "textNoCardChosen2";
    let feedbackNotChosenId = "textNoCardChosen3";
    let triggerChosenId = "trigger";
    let thingChosenId = "thing";
    let feedbackChosenId = "feedback";
    let errorOccured = false; // value turns true if error occured 
    let colorBorder; // color of border and font when no card is picked

    // This function checks which cards are checked off and sets a bool value
    function CheckCards() {
        if ((localStorage.getItem("trigger") === "null" || localStorage.getItem("trigger") === null) && (localStorage.getItem("thing") === "null" || localStorage.getItem("thing") === null) && (localStorage.getItem("feedback") === "null" || localStorage.getItem("feedback") === null)) {
            triggerChosen = false;
            thingChosen = false;
            feedbackChosen = false;
        } else if ((!(localStorage.getItem("trigger") === "null" || localStorage.getItem("trigger") === null)) && (localStorage.getItem("thing") === "null" || localStorage.getItem("thing") === null) && (localStorage.getItem("feedback") === "null" || localStorage.getItem("feedback") === null)) {
            triggerChosen = true;
            thingChosen = false;
            feedbackChosen = false;
        } else if ((!(localStorage.getItem("trigger") === "null" || localStorage.getItem("trigger") === null)) && (!(localStorage.getItem("thing") === "null" || localStorage.getItem("thing") === null)) && (localStorage.getItem("feedback") === "null" || localStorage.getItem("feedback") === null)) {
            triggerChosen = true;
            thingChosen = true;
            feedbackChosen = false;
        } else if ((!(localStorage.getItem("trigger") === "null" || localStorage.getItem("trigger") === null)) && (!(localStorage.getItem("thing") === "null" || localStorage.getItem("thing") === null)) && (!(localStorage.getItem("feedback") === "null" || localStorage.getItem("feedback") === null))) {
            triggerChosen = true;
            thingChosen = true;
            feedbackChosen = true;
        } else {
            // if none of the above are true, then errorOccurs variable is set to true and this is used in the return statement
            errorOccured = true;
        }
    }


    // If bool value is true, then this function is returned. This returns the image and uses the LinkToPage-component
    function trueValue(cardType) {
        if (errorOccured) { // error-solution for now, this will be fixed more professionally later on
            return <LinkButton target="/categories" title={cardType.charAt(0).toUpperCase+cardType.substring(1)} category={cardType} size="Small"></LinkButton>
        } else {
            let cardTypeName = cardType.charAt(0).toUpperCase()+cardType.substring(1)
            return <LinkToPage target="/categories" title={cardTypeName} category={cardType} nameOfCardType={localStorage.getItem(cardType)} classNameCard={"chosen" + cardTypeName}></LinkToPage>
        }
    }

    // If bool value is false, then this function is returned. This function returns a placeholder image and a button
    function falseValue(cardType) {
        if (errorOccured) { // error-solution for now, this will be fixed
            return <LinkButton target="/categories" title={cardType} category={cardType} size="Small"></LinkButton>
        } else {
            // Checks which type of card it is + sets the color of the border
            let classNameOfCard;
            if (cardType === triggerChosenId) {
                classNameOfCard = triggerNotChosenId;
                colorBorder = "#F08A00";
            } else if (cardType === thingChosenId) {
                classNameOfCard = thingNotChosenId;
                colorBorder = "#E53D2F";
            } else if (cardType === feedbackChosenId) {
                classNameOfCard = feedbackNotChosenId;
                colorBorder = "#FFCD00";
            }

            // Returns the placeholder image and button
            // (fjern kommentar: kanskje litt dårlig navngivning på divene og kanskje unødvendig mange, men ble finest og mest responsivt med 2 diver)
            return <div id="firstWrapperCardNotChosen"><div className="wrapperCardNotChosen">
            <img src={placeholderImage} alt="test" className="noImgChosen" style={{ border: `${colorBorder}`, borderStyle: "solid", borderWidth: "0.65vw" }} />
            <div id={classNameOfCard} style={{ color: `${colorBorder}`}}>
                {cardType.charAt(0).toUpperCase() + cardType.substring(1)}
                </div>
            <br/>
            </div>
            <div id="buttonChosenCardsPage">
                 {/* Following code decides whether button is going to be enabled or not */}
            { ((cardType === triggerChosenId) && (!(thingChosen)) && (!(feedbackChosen))) ? <LinkButton target="/categories" title="Choose!" category="trigger" size="chosenCards" className="buttonChosenCardsPage"></LinkButton> : ""}
            { ((cardType === thingChosenId) && (triggerChosen) && (!(feedbackChosen))) ? <LinkButton target="/categories" title="Choose!" category="thing" size="chosenCards" className="buttonChosenCardsPage"></LinkButton> : ""}
            { ((cardType === feedbackChosenId) && (triggerChosen) && (thingChosen)) ? <LinkButton target="/categories" title="Choose!" category="feedback" size="chosenCards" className="buttonChosenCardsPage"></LinkButton> : ""}
            
            </div>
            </div>

            }
        }

    return (
        <div className="Page">
        <TopContainer
            text = "Some text"
        />

            <div className="chosenCardsGrid">
            <div className="chosenCardsGrid2">
            {CheckCards()}
            {/* If-else statements to check the bool-variables */}
            {(triggerChosen) ? <div className="chosenCardContainer">{trueValue(triggerChosenId)}</div> : falseValue(triggerChosenId)}
            {(thingChosen) ? <div className="chosenCardContainer">{trueValue(thingChosenId)}</div> : falseValue(thingChosenId)}
            {(feedbackChosen) ? <div className="chosenCardContainer">{trueValue(feedbackChosenId)}</div> : falseValue(feedbackChosenId)}
            </div>
            </div>



            <br/>
            <LinkButton target="/animation" title="RUN!" category="animation" size="Large"></LinkButton>

        </div>
    );
}

export default ChosenCardsPage