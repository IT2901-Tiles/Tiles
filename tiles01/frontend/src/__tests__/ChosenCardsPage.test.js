import {configure, shallow} from "enzyme"
import Adapter from 'enzyme-adapter-react-16';
import ChosenCardsPage from '../Pages/ChosenCardsPage';
import {cleanup} from '@testing-library/react';
import renderer from 'react-test-renderer';
import {BrowserRouter as Router} from "react-router-dom";


afterEach(() => {
    cleanup();
});

configure({
    adapter: new Adapter()
});

describe('ChosenCards Page', () => {
    const wrapper = shallow(<ChosenCardsPage/>)
    it('renders correctly', () => {
        const tree = renderer.create(<Router><ChosenCardsPage/></Router>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('renders two LinkButton elements', () => {
        expect(wrapper.find('LinkButton')).toHaveLength(2);
    });
    it('renders a CharNormal', () => {
        expect(wrapper.find('CharNormal')).toHaveLength(1);
    });
    it('renders choose-button for trigger, and not things + feedback', () => {
        expect(wrapper.find('.buttonChosenCardsPage').props().category).toEqual('trigger');
        expect(wrapper.find('.buttonChosenCardsPage').props().category).not.toEqual('things');
        expect(wrapper.find('.buttonChosenCardsPage').props().category).not.toEqual('feedback');
    });
    // it('check that choose-button disappears', () => {
    //     const init = {}
    //     const container = shallow(<ChosenCardsPage {...init}/>)
    //     console.log(container.debug())

    //     expect(container.find('.buttonChosenCardsPage').props().category).not.toEqual('trigger');
    // });
});
