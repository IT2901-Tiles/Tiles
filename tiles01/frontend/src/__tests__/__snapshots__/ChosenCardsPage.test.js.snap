// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChosenCards Page renders correctly 1`] = `
<div
  className="Page ChosenCardsPage"
>
  <div
    className="TopContainer"
  >
    <div
      className="Logo"
    >
      <h1>
        TILES
      </h1>
    </div>
    <div
      className="infoBox"
    >
      <p>
        To select a card, click on the "Choose!" button!
      </p>
    </div>
  </div>
  <div
    id="charNormalWrapper"
  >
    <div
      className="CharNormal"
    >
      <svg
        className="svgCharacterNormal"
        viewBox="0 0 165 150"
      >
        <g
          className="whole"
        >
          <g
            className="HeadParts"
          >
            <rect
              fill="grey"
              height="35"
              id="head"
              stroke="black"
              width="55"
              x="60"
              y="36"
            />
            <rect
              fill="grey"
              height="15"
              id="lEar"
              stroke="black"
              width="5"
              x="53"
              y="46"
            />
            <rect
              fill="grey"
              height="15"
              id="rEar"
              stroke="black"
              width="5"
              x="117"
              y="46"
            />
            <rect
              fill="grey"
              height="6"
              id="topHead"
              stroke="black"
              width="15"
              x="80"
              y="28"
            />
            <circle
              cx="87.5"
              cy="61"
              fill="red"
              id="smile"
              r="8"
              stroke="black"
            />
            <rect
              fill="grey"
              height="10"
              id="smileRect"
              width="19"
              x="78"
              y="52"
            />
            <line
              id="topSmile"
              stroke="black"
              x1="79"
              x2="96"
              y1="62"
              y2="62"
            />
          </g>
          <g
            className="BodyParts"
          >
            <rect
              fill="grey"
              height="8"
              id="neck"
              stroke="black"
              width="15"
              x="80"
              y="73"
            />
            <rect
              fill="grey"
              height="35"
              id="body"
              stroke="black"
              width="55"
              x="60"
              y="83"
            />
          </g>
        </g>
        <g
          className="LightBulb"
        >
          <g
            className="lightShine"
          >
            <line
              id="lightStick1"
              stroke="black"
              strokeWidth="2"
              x1="77"
              x2="82"
              y1="4"
              y2="9"
            >
              <animate
                attributeName="y1"
                dur="2s"
                keyTimes="0;0.5;1"
                repeatCount="indefinite"
                values="4;5;4"
              />
              <animate
                attributeName="x1"
                dur="2s"
                keyTimes="0;0.5;1"
                repeatCount="indefinite"
                values="77;79;77"
              />
            </line>
            <line
              id="lightStick2"
              stroke="yellow"
              strokeWidth="2"
              x1="83"
              x2="85"
              y1="2"
              y2="8"
            >
              <animate
                attributeName="y1"
                begin="0.5s"
                dur="2s"
                keyTimes="0;0.5;1"
                repeatCount="indefinite"
                values="2;4;2"
              />
              <animate
                attributeName="x1"
                begin="0.5s"
                dur="2s"
                keyTimes="0;0.5;1"
                repeatCount="indefinite"
                values="83;85;83"
              />
            </line>
            <line
              id="lightStick3"
              stroke="black"
              strokeWidth="2"
              x1="91"
              x2="89"
              y1="2"
              y2="8"
            >
              <animate
                attributeName="y1"
                begin="1s"
                dur="2s"
                keyTimes="0;0.5;1"
                repeatCount="indefinite"
                values="2;4;2"
              />
              <animate
                attributeName="x1"
                begin="1s"
                dur="2s"
                keyTimes="0;0.5;1"
                repeatCount="indefinite"
                values="91;90;91"
              />
            </line>
            <line
              id="lightStick4"
              stroke="yellow"
              strokeWidth="2"
              x1="97"
              x2="93"
              y1="4"
              y2="9"
            >
              <animate
                attributeName="y1"
                begin="1.5s"
                dur="2s"
                keyTimes="0;0.5;1"
                repeatCount="indefinite"
                values="4;6;4"
              />
              <animate
                attributeName="x1"
                begin="1.5s"
                dur="2s"
                keyTimes="0;0.5;1"
                repeatCount="indefinite"
                values="97;96;97"
              />
            </line>
          </g>
          <circle
            cx="87.5"
            cy="15"
            fill="yellow"
            id="lightBulb"
            r="5"
            stroke="black"
          />
          <line
            id="lightStickMain"
            stroke="black"
            strokeWidth="2"
            x1="87.5"
            x2="87.5"
            y1="20"
            y2="28"
          />
          <animateTransform
            attributeName="transform"
            attributeType="XML"
            begin="0s"
            dur="3s"
            keyTimes="0; 0.25; 0.5; 0.75; 1"
            repeatCount="indefinite"
            type="rotate"
            values="0 87.5 28; 5 87.5 28; 0 87.5 28; -5 87.5 28; 0 87.5 28;"
          />
        </g>
        <g
          className="eyes"
        >
          <polygon
            fill="black"
            id="eyeL"
            points="75 44, 68 56, 82 56"
          />
          <polygon
            fill="black"
            id="eyeR"
            points="97 44, 90 56, 104 56"
          />
        </g>
        <g
          className="leftArm"
        >
          <g
            className="leftArmOver"
          >
            <circle
              cx="53.5"
              cy="86"
              fill="grey"
              id="shoulderL"
              r="5"
              stroke="black"
            />
            <rect
              fill="grey"
              height="16"
              id="overarmL"
              stroke="black"
              width="7"
              x="50"
              y="93"
            />
          </g>
          <g
            className="leftArmUnder"
          >
            <circle
              cx="53.5"
              cy="115.5"
              fill="grey"
              id="elbowL"
              r="5"
              stroke="black"
            />
            <rect
              fill="grey"
              height="7"
              id="underarmL"
              stroke="black"
              transform="rotate(20 53.5 115.5)"
              width="16"
              x="31"
              y="113"
            />
            <rect
              fill="grey"
              height="5"
              id="handLL"
              stroke="black"
              transform="rotate(10 53.5 115.5)"
              width="13"
              x="16"
              y="112"
            />
            <rect
              fill="grey"
              height="5"
              id="handLR"
              stroke="black"
              transform="rotate(35 53.5 115.5)"
              width="13"
              x="17"
              y="117"
            />
            <animateTransform
              attributeName="transform"
              attributeType="XML"
              begin="0.75s"
              dur="5s"
              keyTimes="0; 0.20; 0.3; 0.4; 0.5; 0.75; 1"
              repeatCount="indefinite"
              type="rotate"
              values="-20 53.5 115.5;-47 53.5 115.5; -20 53.5 115.5; -47 53.5 115.5; -20 53.5 115.5; -20 53.5 115.5; -20 53.5 115.5;"
            />
          </g>
          <animateTransform
            attributeName="transform"
            attributeType="XML"
            begin="0s"
            dur="5s"
            keyTimes="0; 0.25; 0.5; 0.75; 0.90; 1"
            repeatCount="indefinite"
            type="rotate"
            values="5 53.5 86;90 53.5 86; 90 53.5 86; 90 53.5 86; 5 53.5 86; 5 53.5 86;"
          />
        </g>
        <g
          className="rightArm"
        >
          <circle
            cx="121.5"
            cy="86"
            fill="grey"
            id="shoulderR"
            r="5"
            stroke="black"
          />
          <rect
            fill="grey"
            height="16"
            id="overarmR"
            stroke="black"
            width="7"
            x="118"
            y="93"
          />
          <circle
            cx="121.5"
            cy="115.5"
            fill="grey"
            id="elbowR"
            r="5"
            stroke="black"
          />
          <rect
            fill="grey"
            height="7"
            id="underarmR"
            stroke="black"
            transform="rotate(-20 121.5 115.5)"
            width="16"
            x="128"
            y="113"
          />
          <rect
            fill="grey"
            height="5"
            id="handRL"
            stroke="black"
            transform="rotate(-30 140 101)"
            width="13"
            x="140"
            y="103"
          />
          <rect
            fill="grey"
            height="5"
            id="handRR"
            stroke="black"
            transform="rotate(10 145 107)"
            width="13"
            x="145"
            y="107"
          />
        </g>
      </svg>
    </div>
  </div>
  <div
    className="chosenCardsGrid"
  >
    <div
      className="chosenCardsGrid2"
    >
      <div
        id="firstWrapperCardNotChosen"
      >
        <div
          className="wrapperCardNotChosen"
        >
          <img
            alt="test"
            className="noImgChosen"
            src="blankImg.png"
            style={
              Object {
                "border": "#F08A00",
                "borderStyle": "solid",
                "borderWidth": "0.65vw",
              }
            }
          />
          <div
            id="noCardChosenTrigger"
            style={
              Object {
                "color": "#F08A00",
              }
            }
          >
            Trigger
          </div>
          <br />
        </div>
        <div
          id="buttonChosenCardsPage"
        >
          <a
            href="/categories"
            onClick={[Function]}
          >
            <button
              className="chosenCardsButton btn btn-primary"
              disabled={false}
              style={
                Object {
                  "backgroundColor": "#F08A00",
                  "color": "#F3EAC2",
                }
              }
              type="button"
            >
              Choose!
            </button>
          </a>
          
          
        </div>
      </div>
      <div
        id="firstWrapperCardNotChosen"
      >
        <div
          className="wrapperCardNotChosen"
        >
          <img
            alt="test"
            className="noImgChosen"
            src="blankImg.png"
            style={
              Object {
                "border": "#E53D2F",
                "borderStyle": "solid",
                "borderWidth": "0.65vw",
              }
            }
          />
          <div
            id="noCardChosenThings"
            style={
              Object {
                "color": "#E53D2F",
              }
            }
          >
            Things
          </div>
          <br />
        </div>
        <div
          id="buttonChosenCardsPage"
        >
          
          
          
        </div>
      </div>
      <div
        id="firstWrapperCardNotChosen"
      >
        <div
          className="wrapperCardNotChosen"
        >
          <img
            alt="test"
            className="noImgChosen"
            src="blankImg.png"
            style={
              Object {
                "border": "#FFCD00",
                "borderStyle": "solid",
                "borderWidth": "0.65vw",
              }
            }
          />
          <div
            id="noCardChosenFeedback"
            style={
              Object {
                "color": "#FFCD00",
              }
            }
          >
            Feedback
          </div>
          <br />
        </div>
        <div
          id="buttonChosenCardsPage"
        >
          
          
          
        </div>
      </div>
    </div>
  </div>
  <br />
  <a
    href="/animation"
    onClick={[Function]}
  >
    <button
      className="LargeButton btn btn-primary"
      disabled={true}
      style={
        Object {
          "backgroundColor": "#84AD64",
          "color": "#F3EAC2",
        }
      }
      type="button"
    >
      RUN!
    </button>
  </a>
</div>
`;
